[CASSANDRA-20190][trunk] MemoryUtil.setInt/getInt and similar use the wrong endianness

netudima opened the pull request:
Add NativeEndianMemoryUtil, move all methods which uses (or should) native endian
Switch offheap memtable primitive fields serialization/deserialization to NativeEndianMemoryUtil

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20190

blambov commented on src/java/org/apache/cassandra/db/rows/NativeCell.java:
```diff
@@ -194,20 +195,20 @@ public long unsharedHeapSizeExcludingData()
     private boolean hasPath()
     {
-        return MemoryUtil.getByte(peer+ HAS_CELLPATH) != 0;
+        return MemoryUtil.getByte(peer + HAS_CELLPATH) != 0;
```
Nit: this should also use `NativeEndianMemoryUtil` for consistency.

netudima replied:
I was not sure here, for one side - yes, it is better for consistency point of view, for another side I wanted to highlight that byte operations are endian agnostic..  I will switch to NativeEndianMemoryUtil usage (in any case I already replaced some similar places)

blambov commented on test/unit/org/apache/cassandra/utils/memory/NativeEndianMemoryUtilTest.java:
```diff
@@ -0,0 +1,100 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+public class NativeEndianMemoryUtilTest
```
We need to also test the `byByte` versions; perhaps make them package private and call directly here.

netudima replied:
yes, I have the same plan - to invoke them directly or to play with arch system property in a unit test

netudima replied:
I have added tests via direct invocations of package private methods

netudima committed:
Add LittleEndianMemoryUtil and NativeEndianMemoryUtil, switch memtable-related off-heap objects and Memory to use them and have Little Endian now.
Add BE offsets detection on Summary loading.
Add test SSTables in an old format with BE offsets in Summary component to LegacySSTableTest.

Patch by Dmitry Konstantinov; reviewed by Branimir Lambov, Michael Semb Wever for CASSANDRA-20190

blambov commented on src/java/org/apache/cassandra/utils/memory/LittleEndianMemoryUtil.java:
```diff
@@ -0,0 +1,146 @@
+
+    public static void setLong(long address, long l)
+    {
+        if (Architecture.IS_UNALIGNED)
```
Shouldn't we have the aligned address option (e.g. `|| (address & 0b111) == 0L`) to use a direct store in the writes as well?

netudima replied:
thank you for noticing the gap, done

netudima commented:
rebased and squashed
